dlt/destinations/impl/sink/sink.py

```diff
@@ -0,0 +1,172 @@
+        self._storage_id = f"{self._parsed_file_name.table_name}.{self._parsed_file_name.file_id}"
+        try:
+            current_index = load_package_state.get(self._storage_id, 0)
+            for batch in self.run(current_index):
+                self.call_callable_with_items(batch)
+                current_index += len(batch)
+                load_package_state[self._storage_id] = current_index
+
+            self._state = "completed"
+        except Exception as e:
+            self._state = "retry"
+            raise e
+        finally:
+            # save progress
+            commit_load_package_state()
+
+    @abstractmethod
+    def run(self, start_index: int) -> Iterable[TDataItems]:
+        pass
+
+    def call_callable_with_items(self, items: TDataItems) -> None:
+        if not items:
+            return
+
+        # coerce items into correct format specified by schema
+        coerced_items: TDataItems = []
+        for item in items:
+            coerced_item, table_update = self._schema.coerce_row(self._table["name"], None, item)
+            assert not table_update
+            coerced_items.append(coerced_item)
+
+        # send single item on batch size 1
+        if self._config.batch_size == 1:
+            coerced_items = coerced_items[0]
+
+        # call callable
+        self._config.credentials.resolved_callable(coerced_items, self._table)
+
+    def state(self) -> TLoadJobState:
+        return self._state
+
+    def exception(self) -> str:
+        raise NotImplementedError()
+
+
+class SinkParquetLoadJob(SinkLoadJob):
+    def run(self, start_index: int) -> Iterable[TDataItems]:
+        # stream items
+        from dlt.common.libs.pyarrow import pyarrow
+
+        # guard against changed batch size after restart of loadjob
+        assert (
+            start_index % self._config.batch_size
+        ) == 0, "Batch size was changed during processing of one load package"
+
+        start_batch = start_index / self._config.batch_size
+        with pyarrow.parquet.ParquetFile(self._file_path) as reader:
+            for record_batch in reader.iter_batches(batch_size=self._config.batch_size):
+                if start_batch > 0:
+                    start_batch -= 1
+                    continue
+                batch = record_batch.to_pylist()
+                yield batch
+
+
+class SinkJsonlLoadJob(SinkLoadJob):
+    def run(self, start_index: int) -> Iterable[TDataItems]:
+        current_batch: TDataItems = []
+
+        # stream items
+        with FileStorage.open_zipsafe_ro(self._file_path) as f:
+            for line in f:
+                # find correct start position
+                if start_index > 0:
+                    start_index -= 1
+                    continue
+                current_batch.append(json.loads(line))
+                if len(current_batch) == self._config.batch_size:
+                    yield current_batch
+                    current_batch = []
+            yield current_batch
```

Comment:
should we check if batch actually has data?

Reply:
this is already done in the call_callable_with_items function above